lib/src/execution/posix/execution.cc

```diff
@@ -176,6 +178,131 @@ namespace facter { namespace execution {
         throw timeout_exception((boost::format("command timed out after %1% seconds.") % timeout).str(), static_cast<size_t>(child));
     }
 
+    static void exec_child(int in, int out, int err, char const* program, char const** argv, char const** envp)
+    {
+        // WARNING: this function is called from a vfork'd child
+        // Do not modify program state from this function; only call setpgid, dup2, close, execve, and _exit
+        // Do not allocate heap memory or throw exceptions
+        // The child is sharing the address space of the parent process, so carelessly modifying this
+        // function may lead to parent state corruption, memory leaks, and/or total protonic reversal
+
+        // Set the process group; this will be used by the parent if we need to kill the process and its children
+        if (setpgid(0, 0) == -1) {
+            char const* message = "failed to setpgid.";
+            if (write(err, message, strlen(message)) == -1) {
+                // Do not care
+            }
+            return;
+        }
+
+        // Redirect stdin
+        if (dup2(in, STDIN_FILENO) == -1) {
+            char const* message = "failed to redirect child stdin.";
+            if (write(err, message, strlen(message)) == -1) {
+                // Do not care
+            }
+            return;
+        }
+
+        // Redirect stdout
+        if (dup2(out, STDOUT_FILENO) == -1) {
+            char const* message = "failed to redirect child stdout.";
+            if (write(err, message, strlen(message)) == -1) {
+                // Do not care
+            }
+            return;
+        }
+
+        // Redirect stderr
+        if (dup2(err, STDERR_FILENO) == -1) {
+            char const* message = "failed to redirect child stderr.";
+            if (write(err, message, strlen(message)) == -1) {
+                // Do not care
+            }
+            return;
+        }
+
+        // Close all open file descriptors above stderr
+        for (decltype(get_max_descriptor_limit()) i = (STDERR_FILENO + 1); i < get_max_descriptor_limit(); ++i) {
+            close(i);
+        }
+
+        // Execute the given program; this should not return if successful
+        execve(program, const_cast<char* const*>(argv), const_cast<char* const*>(envp));
+    }
+
+    // Helper function that turns a vector of strings into a vector of const cstr pointers
+    // This is used to pass arguments and environment to execve
+    static vector<char const*> to_exec_arg(vector<string> const* argument, string const* first = nullptr)
+    {
+        vector<char const*> result((argument ? argument->size() : 0) + (first ? 1 : 0) + 1 /* terminating null */);
```

Comment:
For some reason I have a preference for: reserve, push_back first, transform with back_inserter, push_back null. But they do pretty much identical things.